geoh5py/objects/vp_model.py

```diff
@@ -0,0 +1,380 @@
+
+    @heterogeneous_property_id.setter
+    def heterogeneous_property_id(self, value: uuid.UUID):
+        if not isinstance(value, uuid.UUID):
+            raise TypeError(
+                "Attribute 'heterogeneous_property_id' should be a 'uuid.UUID'."
+            )
+
+        self._heterogeneous_property_id = value
+
+        if self.on_file:
+            self.workspace.update_attribute(self, "attributes")
+
+    @property
+    def physical_data_name(self) -> str:
+        """
+        VPmg physical property model data map name.
+        """
+        return self._physical_data_name
+
+    @physical_data_name.setter
+    def physical_data_name(self, value: str):
+        if not isinstance(value, str):
+            raise TypeError("Attribute 'physical_data_name' should be a 'str'")
+
+        self._physical_data_name = value
+
+        if self.on_file:
+            self.workspace.update_attribute(self, "attributes")
+
+    @property
+    def unit_property_id(self) -> uuid.UUID | str | None:
+        """
+        VPmg unit model id.
+        """
+        return self._unit_property_id
+
+    @unit_property_id.setter
+    def unit_property_id(self, value: uuid.UUID):
+        if not isinstance(value, uuid.UUID):
+            raise TypeError("Attribute 'unit_property_id' should be a 'uuid.UUID'.")
+
+        self._unit_property_id = value
+
+        if self.on_file:
+            self.workspace.update_attribute(self, "attributes")
+
+    @property
+    def weight_property_id(self) -> uuid.UUID | str | None:
+        """
+        VPmg weight model id.
+        """
+        return self._weight_property_id
+
+    @weight_property_id.setter
+    def weight_property_id(self, value: uuid.UUID):
+        if not isinstance(value, uuid.UUID):
+            raise TypeError(
+                "Attribute 'weight_property_id' should be a 'uuid.UUID' or None"
+            )
+        self._weight_property_id = value
+
+        if self.on_file:
+            self.workspace.update_attribute(self, "attributes")
+
+    @property
+    def centroids(self) -> np.ndarray:
+        """
+        Cell center locations in world coordinates, shape(*, 3).
+
+        .. code-block:: python
+
+            centroids = [
+                [x_1, y_1, z_1],
+                ...,
+                [x_N, y_N, z_N]
+            ]
+        """
+        if getattr(self, "_centroids", None) is None:
+            rotation_matrix = xy_rotation_matrix(np.deg2rad(self.rotation))
+
+            v_grid, u_grid = np.meshgrid(
+                np.cumsum(np.ones(self._v_count) * self.v_cell_size)
+                - self.v_cell_size / 2,
+                np.cumsum(np.ones(self._u_count) * self.u_cell_size)
+                - self.u_cell_size / 2,
+            )
+            xyz = (
+                rotation_matrix
+                @ np.c_[np.ravel(u_grid), np.ravel(v_grid), self.prisms[:, 0]].T
+            ).T
+
+            for ind, axis in enumerate(["x", "y", "z"]):
```

Comment:
not the best those recarray for operation?
Not a big deal but do you want to change this as linear operation converting the recarray to np array?

Reply:
Yes, I will convert in place though. Otherwise we need to do something more complicated when writing out to the file.